// @sap-ux/eslint-plugin-fiori-tools/lib/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.rules = exports.configs = exports.languages = exports.meta = exports.DiagnosticCache = void 0;
const node_fs_1 = require("node:fs");
const node_path_1 = require("node:path");
const js_1 = __importDefault(require("@eslint/js"));
const eslint_parser_1 = __importDefault(require("@babel/eslint-parser"));
const eslint_plugin_1 = __importDefault(require("@typescript-eslint/eslint-plugin"));
const rules_1 = require("./rules");
const fiori_language_1 = require("./language/fiori-language");
const synckit_1 = require("synckit");
const utils_1 = require("./project-context/utils");
var diagnostic_cache_1 = require("./language/diagnostic-cache");
Object.defineProperty(exports, "DiagnosticCache", { enumerable: true, get: function () { return diagnostic_cache_1.DiagnosticCache; } });
// Use CommonJS require for modules with resolution issues
// eslint-disable-next-line @typescript-eslint/no-require-imports
const tsParser = require('@typescript-eslint/parser');
// eslint-disable-next-line @typescript-eslint/no-require-imports
const globals = require('globals');
// Read package.json to get version
const packageJson = JSON.parse((0, node_fs_1.readFileSync)((0, node_path_1.join)(__dirname, '../package.json'), 'utf-8'));
/**
 * Plugin meta information (required for ESLint 9).
 * Contains the plugin name and version.
 */
exports.meta = {
    name: packageJson.name,
    version: packageJson.version
};
/**
 * Language definitions supported by the plugin.
 * Currently includes the Fiori language for annotation and manifest files.
 */
exports.languages = {
    fiori: new fiori_language_1.FioriLanguage()
};
/**
 * Default export following ESLint 9 plugin structure.
 * This is the recommended way to export plugins in ESLint 9.
 * Contains plugin metadata, supported languages, rules, and processors.
 */
const plugin = {
    meta: {
        name: packageJson.name,
        version: '0.0.1',
        namespace: '@sap-ux/fiori-tools'
    },
    languages: exports.languages,
    rules: rules_1.rules,
    processors: {}
};
/**
 * Common configuration shared across all config presets.
 */
const commonConfig = [
    {
        languageOptions: {
            globals: {
                ...globals.browser,
                ...globals.node,
                sap: 'readonly'
            }
        }
    }
];
// Use synckit to create sync function for project-access getPathMappingsSync
const workerPath = (0, node_path_1.join)(__dirname, 'worker-getPathMappingsSync.js');
const getPathMappingsSync = (0, synckit_1.createSyncFn)(workerPath);
const pathMappingsAbsolute = getPathMappingsSync(process.cwd());
const webappPathAbsolute = 'webapp' in pathMappingsAbsolute
    ? pathMappingsAbsolute.webapp
    : (pathMappingsAbsolute.src ?? (0, node_path_1.join)(process.cwd(), 'webapp'));
const webappPathRelative = (0, utils_1.uniformUrl)((0, node_path_1.relative)(process.cwd(), webappPathAbsolute));
const testPathRelative = 'webapp' in pathMappingsAbsolute
    ? node_path_1.posix.join(webappPathRelative, 'test')
    : (0, utils_1.uniformUrl)((0, node_path_1.relative)(process.cwd(), pathMappingsAbsolute.test ?? (0, node_path_1.join)(process.cwd(), 'webapp/test')));
const prodConfig = [
    {
        files: [`./${webappPathRelative}/**/*.js`, `./${webappPathRelative}/**/*.ts`],
        ignores: [
            'target/**',
            `${testPathRelative}/**`,
            `${node_path_1.posix.join(webappPathRelative, 'localservice')}/**`, // Ignore everything in the 'localservice' folder
            `!${node_path_1.posix.join(webappPathRelative, 'localservice')}/**/*.{ts,js}`, // EXCEPT for .ts and .js files (that might be custom mockserver extensions)
            `${node_path_1.posix.join(webappPathRelative, 'localService')}/**`, // Ignore everything in the 'localService' folder
            `!${node_path_1.posix.join(webappPathRelative, 'localService')}/**/*.{ts,js}`, // EXCEPT for .ts and .js files (that might be custom mockserver extensions)
            'backup/**',
            '**/Gruntfile.js',
            '**/changes_preview.js',
            '**/changes_preview.ts',
            '**/gulpfile.js',
            '**/*.d.ts',
            'test/**'
        ],
        languageOptions: {
            parser: eslint_parser_1.default,
            parserOptions: {
                requireConfigFile: false
            }
        },
        rules: {
            ...js_1.default.configs.recommended.rules,
            // Error rules (alphabetical)
            '@sap-ux/fiori-tools/sap-no-absolute-component-path': 'error',
            '@sap-ux/fiori-tools/sap-no-br-on-return': 'error',
            '@sap-ux/fiori-tools/sap-no-commons-usage': 'error',
            '@sap-ux/fiori-tools/sap-no-dynamic-style-insertion': 'error',
            '@sap-ux/fiori-tools/sap-no-element-creation': 'error',
            '@sap-ux/fiori-tools/sap-no-exec-command': 'error',
            '@sap-ux/fiori-tools/sap-no-global-define': 'error',
            '@sap-ux/fiori-tools/sap-no-global-event': 'error',
            '@sap-ux/fiori-tools/sap-no-global-variable': 'error',
            '@sap-ux/fiori-tools/sap-no-hardcoded-color': 'error',
            '@sap-ux/fiori-tools/sap-no-hardcoded-url': 'error',
            '@sap-ux/fiori-tools/sap-no-inner-html-write': 'error',
            '@sap-ux/fiori-tools/sap-no-localstorage': 'error',
            '@sap-ux/fiori-tools/sap-no-location-reload': 'error',
            '@sap-ux/fiori-tools/sap-no-navigator': 'error',
            '@sap-ux/fiori-tools/sap-no-override-rendering': 'error',
            '@sap-ux/fiori-tools/sap-no-override-storage-prototype': 'error',
            '@sap-ux/fiori-tools/sap-no-sessionstorage': 'error',
            '@sap-ux/fiori-tools/sap-no-ui5base-prop': 'error',
            // Warning rules (alphabetical)
            '@sap-ux/fiori-tools/sap-bookmark-performance': 'warn',
            '@sap-ux/fiori-tools/sap-browser-api-warning': 'warn',
            '@sap-ux/fiori-tools/sap-cross-application-navigation': 'warn',
            '@sap-ux/fiori-tools/sap-forbidden-window-property': 'warn',
            '@sap-ux/fiori-tools/sap-message-toast': 'warn',
            '@sap-ux/fiori-tools/sap-no-dom-access': 'warn',
            '@sap-ux/fiori-tools/sap-no-dom-insertion': 'warn',
            '@sap-ux/fiori-tools/sap-no-encode-file-service': 'warn',
            '@sap-ux/fiori-tools/sap-no-global-selection': 'warn',
            '@sap-ux/fiori-tools/sap-no-history-manipulation': 'warn',
            '@sap-ux/fiori-tools/sap-no-inner-html-access': 'warn',
            '@sap-ux/fiori-tools/sap-no-jquery-device-api': 'warn',
            '@sap-ux/fiori-tools/sap-no-localhost': 'warn',
            '@sap-ux/fiori-tools/sap-no-location-usage': 'warn',
            '@sap-ux/fiori-tools/sap-no-proprietary-browser-api': 'warn',
            '@sap-ux/fiori-tools/sap-no-ui5-prop-warning': 'warn',
            '@sap-ux/fiori-tools/sap-timeout-usage': 'warn',
            '@sap-ux/fiori-tools/sap-ui5-forms': 'warn',
            '@sap-ux/fiori-tools/sap-ui5-global-eval': 'warn',
            '@sap-ux/fiori-tools/sap-ui5-legacy-factories': 'warn',
            '@sap-ux/fiori-tools/sap-ui5-legacy-jquerysap-usage': 'warn',
            '@sap-ux/fiori-tools/sap-usage-basemastercontroller': 'warn',
            // Off rules (alphabetical)
            '@sap-ux/fiori-tools/sap-browser-api-error': 'off',
            '@sap-ux/fiori-tools/sap-no-window-alert': 'off',
            '@sap-ux/fiori-tools/sap-ui5-no-private-prop': 'off'
        }
    }
];
const testConfig = [
    {
        files: [`./${testPathRelative}/**/*.js`, `./${testPathRelative}/**/*.ts`],
        ignores: ['**/*.d.ts'],
        languageOptions: {
            parser: eslint_parser_1.default,
            parserOptions: {
                requireConfigFile: false
            }
        },
        rules: {
            '@sap-ux/fiori-tools/sap-opa5-autowait-true': 'warn'
        }
    }
];
const typescriptConfig = [
    {
        files: [`./${webappPathRelative}/*.ts`, `./${webappPathRelative}/**/*.ts`],
        ignores: [
            'target/**',
            `${testPathRelative}/changes_loader.ts`,
            `${testPathRelative}/changes_preview.ts`,
            `${node_path_1.posix.join(webappPathRelative, 'localservice')}/**`, // Ignore everything in the 'localservice' folder
            `!${node_path_1.posix.join(webappPathRelative, 'localservice')}/**/*.{ts,js}`, // EXCEPT for .ts and .js files (that might be custom mockserver extensions)
            `${node_path_1.posix.join(webappPathRelative, 'localService')}/**`, // Ignore everything in the 'localService' folder
            `!${node_path_1.posix.join(webappPathRelative, 'localService')}/**/*.{ts,js}`, // EXCEPT for .ts and .js files (that might be custom mockserver extensions)
            'undefined/**/Example.qunit.ts',
            'backup/**',
            '**/*.d.ts'
        ],
        plugins: {
            '@typescript-eslint': eslint_plugin_1.default
        },
        languageOptions: {
            parser: tsParser,
            ecmaVersion: 5,
            sourceType: 'script',
            parserOptions: {
                projectService: true
            }
        },
        rules: {
            ...eslint_plugin_1.default.configs.recommended.rules,
            ...eslint_plugin_1.default.configs['recommended-type-checked'].rules,
            // Warning rules (alphabetical)
            '@typescript-eslint/no-unsafe-argument': 'warn',
            '@typescript-eslint/no-unsafe-assignment': 'warn',
            '@typescript-eslint/no-unsafe-call': 'warn',
            '@typescript-eslint/no-unsafe-member-access': 'warn',
            '@typescript-eslint/no-unsafe-return': 'warn'
        }
    }
];
// Named configs for easy consumption
exports.configs = {
    // Recommended config for JavaScript & TypeScript projects (prod + test)
    recommended: [
        {
            plugins: {
                '@sap-ux/fiori-tools': {
                    meta: exports.meta,
                    rules: rules_1.rules
                }
            }
        },
        ...commonConfig,
        ...typescriptConfig,
        ...prodConfig,
        ...testConfig
    ],
    // Recommended for S/4HANA config for JavaScript & TypeScript projects
    'recommended-for-s4hana': [
        {
            plugins: {
                '@sap-ux/fiori-tools': {
                    meta: exports.meta,
                    languages: exports.languages,
                    rules: rules_1.rules
                }
            }
        },
        ...commonConfig,
        ...typescriptConfig,
        ...prodConfig,
        ...testConfig,
        {
            files: ['**/manifest.json', '**/*.xml', '**/*.cds'],
            language: '@sap-ux/fiori-tools/fiori',
            rules: {
                '@sap-ux/fiori-tools/sap-anchor-bar-visible': 'warn',
                '@sap-ux/fiori-tools/sap-condensed-table-layout': 'warn',
                '@sap-ux/fiori-tools/sap-flex-enabled': 'warn',
                '@sap-ux/fiori-tools/sap-width-including-column-header': 'warn',
                '@sap-ux/fiori-tools/sap-copy-to-clipboard': 'warn',
                '@sap-ux/fiori-tools/sap-enable-export': 'warn',
                '@sap-ux/fiori-tools/sap-enable-paste': 'warn',
                '@sap-ux/fiori-tools/sap-creation-mode-for-table': 'warn',
                '@sap-ux/fiori-tools/sap-state-preservation-mode': 'warn',
                '@sap-ux/fiori-tools/sap-strict-uom-filtering': 'warn',
                '@sap-ux/fiori-tools/sap-table-personalization': 'warn',
                '@sap-ux/fiori-tools/sap-table-column-vertical-alignment': 'warn'
            }
        }
    ]
};
var rules_2 = require("./rules");
Object.defineProperty(exports, "rules", { enumerable: true, get: function () { return rules_2.rules; } });
exports.default = plugin;
